### DISCUSSION TITLE
How to test Directus flows? What are you using flows for?

### DISCUSSION THREAD

**Discussion by tofran** [directus/directus#15205] [Archive]
I have a dilemma, should one try to build complex logic with Directus flows or just use custom coded extensions?

At the same time, the slogan for flows is:
> Automate Complex Data Processes for Your Digital Projects

The problem is that creating flows is something quite technical, and is not easily auditable since the output is stored in the database/migrations (vs code in git).
I also could not find a way to predictable test a flow. The only thing I managed to do is try to replicate the trigger by hand, every time I want to test if the flow works. For example, creating an entity over an over again, by hand. You might say I can convert that to a cURL request, the problem is that for more complex situations (ex where there is some constraint) you have to setup/delete a bunch of entities every time. You can eventually develop a _fixture like_ script that does that, but this sounds like a rabbit hole.

What are your uses for flows?
What is the best, recommended way for creating quality testable logic on top of Directus?

**Comment by barakplasma** [comment]
We are using flows a bit. We have a complex DB ERD where adding one row in table A requires a related row to be added in table B,C,D etc. So I added a flow that creates the foreign key related rows in the other nested tables.

- Table A: new row
    - Table B: triggered by A
        - Table C: triggered by B

I have no idea how to test it either, so I subscribed to this discussion too. The cURL black box test option sounds as good to me as anything else. ( [basically same as testing a DB trigger](https://www.softwaretestinghelp.com/database-testing-process/) )

**Comment by rijkvanzanten** [comment]
@erondpowell 🙂

**Reply by erondpowell** [reply]
@tofran @barakplasma 

### Great questions and excellent timing!
We are about to completely overhaul and expand the `Flows` docs and also write a "Flows Cookbook".
Hopefully we can build momentum on this discussion and channel any insights into the guides and docs.

### Use Cases
> Should one try to build complex logic with Directus flows or just use custom coded extensions?

Flows can do a lot of stuff and its a totally open-ended tooling system! That said, it is also a very new tooling system, and any improvement suggestions are welcome! 

Choosing between extensions vs flows will depend on several factors, including the team's skills & team preference. First, flows definitely make things more accessible. A business user or SQL analyst may not be able to write their own scripts or Extensions.... but they could learn to use flows fairly quickly. Second, flows also keep your Directus project *lean* as custom Extensions may add more complexity when managing/updating your project.

### Testability
> ... creating an entity over and over again, by hand... You can eventually develop a [cURL] *fixture like* script that does that

At first, I was creating items manually in the app... This was super tedious, but worked ok for simple flows. Eventually, I started using Postman to test event triggers via the API, which made things a bit easier as you get a nice GUI and workspace that can be configured and saved for future use. But IDK if Postman beats cURL blackbox testing! 😃 tbh I haven't explicitly used cURL like that really.

> you have to setup/delete a bunch of entities every time.

When batch deleting data with the APP or API, make sure your dummy data has some value to query on and easily delete.

> What is the best, recommended way for creating quality testable logic on top of Directus?

For flows, at the current stage, you'll setting up logs liberally, triggering the flows via APP/API, and then making sure the test data is easy to query and delete. 

Also, be sure you have a development and production instance.

**Reply by direnc** [reply]
Since you are writing docs for Flows, I would definitely recommend providing sample code for different use cases. For instance it would be helpful if you provide an example for updating a specific field in the condition of updating some other field. Accessing the variables does not feel intuitive for me, not sure either what the "Keys" are for in the flows.

**Reply by emilyf** [reply]
I really like the idea of the cookbook and sample use cases. It's certainly a large task to take on and will need contributors (I would love to participate/contribute to this!!!), but here's some crazy ideas of dream scenarios:
- sandbox from the docs to example read only flows ??? Dang that would be cool. By read only I mean ability to click into the flow and view the operation details/logs etc. 
- Or even cooler would be equivalent of something a stack blitz of pre-configured flows that you can go into and edit/test/modify (sorry, I know stackblitz is not the right example here since we are talking interface and not code but it was closest thing I could come up with)
- Links to video examples that go through in detail example flows/operations etc
- All options that come with `{{ $trigger}}' and `{{ $last }}` .... for example, difference between `{{ $last.my_field }}`, `{{ $last.body.keys }}`, `{{ $trigger.body.collection }}`, `{{ $trigger.payload.my_field}}` etc
- autocomplete on queries and payloads based on your own data model!!!!!! Aka equiv of code autocomplete type approach upon package import etc (dream list)]
- a "demo/tutorial" approach when you first go into flows (this one is least interesting to me and i think could be achieved instead through something like flows that ship with your install and more help text etc
- debugger / logger mentioned on https://github.com/directus/directus/discussions/15870#discussioncomment-3823606 would be AWESOME
- https://github.com/directus/directus/discussions/14010#discussioncomment-3002473

I'd love to know which ones of the above appeal to Directus team and seem most within reach, and I'll sign myself up to help in any way I can!

**Reply by erondpowell** [reply]
@direnc Yeah good suggestions! I think those "sample code" snippets will go into something like the Flows Cookbook. 
That said, we're looking in to replacing the images in flows docs with videos to explicitly walk through the options for each trigger and operation.

>  (I would love to participate/contribute to this!!!)

Yeeeeehaw! 🤠 @emilyf 

Interestingly, your whole flows-playground/StackBlitz idea is actually in the brain-storming stages! (CC @freekrai )

All your ideas are really good. The key is rolling out these ideas in a clean, efficient and scalable way.
And then balancing that with the ⛰️  of other docs-land stuff we got cooking.

**Reply by erondpowell** [reply]
@emilyf P.S. If you ever want to setup time for a video call, I'd love to hear about your project(s) and how you're using flows, hear about your victories and pain-points getting it all setup, and try to extract what lessons I can for the all those who visit docs-land. lmk if interested!

**Reply by direnc** [reply]
I would also be interested in contribution, providing feedback etc. and discussion if needed.

**Reply by erondpowell** [reply]
> I would also be interested in contribution, providing feedback etc. and discussion if needed.

@direnc Let's do it! Find me on Discord ( eron_powell ) and we can setup a time from there.

**Comment by haniefss** [comment]
> ... should one try to build complex logic with Directus flows or just use custom coded extensions?

@erondpowell Since each and every `operation` is stored in db separately, is there gonna be a (significant) performance loss when choosing the flows path (instead of extensions, for the same tasks)? Performance hit raises as the task get more complex, am I wrong to think of it this way?

**Reply by rijkvanzanten** [reply]
That's luckily not really how that works 🙂 The operations of the flow are all read from the DB at once, and then just executed sequentially, so setting up a three-step flow is really no different from writing a hook that does:

```js
await sendRequest();
await sendEmail();
await saveToDatabase();
```

**Reply by rijkvanzanten** [reply]
So yeah, there's _some_ uses where you could do things in parallel, but for most use cases the perf difference should be negligible, seeing that more often than not you'd use the result from step 1 in step 2, etc
